### PR TITLE
twitter classifcation / Retriever 코드 폴더 위치 변경

### PR DESCRIPTION
**Pull request by kimbyeolhee**
## [2023-01-27] twitter classifcation 및 Retriever 코드 폴더 위치 변경
- merge 요청자: @kimbyeolhee

## 1. PR 목적
- 폴더 구조 개편에 맞추어 모듈마다의 동작을 구분하기 위함

## 2. PR 내용
- twitter_classification
    - 기존에는 data/twitter_classification에 위치했으나 corpus 폴더 하단으로 이동
- Retriever
    - chatbot 폴더 하단으로 이동
    - dpr의 경우 사용하지 않을 가능성이 크지만 학습 코드는 chatbot/retriever/dpr에 위치
    - 사용하게 된다면 inference 모듈은 chatbot/retriever에 dense_retriever.py로 구현 예정

### CODE REVIEW

File: chatbot/retriever/dpr/dpr_dataset.py [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+        query = self.queries[idx]
+        context = self.contexts[idx]
+
+        query_tensor = self.tokenize(query)
+        context_tensor = self.tokenize(context)
+
+        return (
+            query_tensor["input_ids"],
+            query_tensor["attention_mask"],
+            query_tensor["token_type_ids"],
```
Comment (wjlee-ling) [inline chatbot/retriever/dpr/dpr_dataset.py:24]: 나중에 DPR를 훈련시킨다고 하면 dictionary꼴로 return하는 걸로 바꿔야 할 것 같아요. token_type_ids가 필요 없는 모델들도 있으니까요..

File: chatbot/retriever/dpr/dpr_trainer.py [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        logger.addHandler(file_handler)
+
+    def configure_optimizer(self, optimizer_grouped_parameters, config):
+        optimizer = torch.optim.AdamW(optimizer_grouped_parameters, lr=config.trainer.learning_rate)
+        return optimizer
+
+    def train_per_epoch(self, epoch_iterator: DataLoader, optimizer):
+        batch_loss = 0
+
+        for step, batch in enumerate(epoch_iterator):
```
Comment (wjlee-ling) [inline chatbot/retriever/dpr/dpr_trainer.py:54]: 여기도 나중에 dpr 사용하게 되면 tqdm 도입해야 할 것같아요. validation에는 있는데 train에는 없네요

File: chatbot/retriever/elastic_retriever.py [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+        with open("./chatbot/retriever/setting.json", "r") as f:
+            setting = json.load(f)
+
+        self.index_name = "chatbot"
+        if self.es.indices.exists(index=self.index_name):
+            self.es.indices.delete(index=self.index_name)
+        self.es.indices.create(index=self.index_name, body=setting)
+
+        # load data
+        if not os.path.exists("./chatbot/retriever/data/elastic_data_v1.json"):  # TODO : 나중에 yaml 파일에서 데이터 경로 받아오도록 수정
```
Comment (wjlee-ling) [inline chatbot/retriever/elastic_retriever.py:40]: elastic_data 는 어떻게 버전 관리 하시는 건가요?

File: chatbot/retriever/elastic_retriever.py [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+        questions = [hit["_source"]["question"] for hit in res["hits"]["hits"]]
+        answers = [hit["_source"]["answer"] for hit in res["hits"]["hits"]]
+        return scores, questions, answers
+
+
+if __name__ == "__main__":
+    elastic_retriever = ElasticRetriever()
+
+    # test
+    query = "태형아 나랑 결혼하자"
```
Comment (wjlee-ling) [inline chatbot/retriever/elastic_retriever.py:81]: to-do Terminal로 인풋 받기

File: corpus/twitter_classification/inference.py [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+if __name__ == "__main__":
+    parser = argparse.ArgumentParser()
+
+    parser.add_argument("--config", "-c", type=str, default="tweet_classification_config")
+
+    args, _ = parser.parse_known_args()
+    config = OmegaConf.load(f"./data/twitter_data_preprocess/{args.config}.yaml")
+
+    # fix random seeds for reproducibility
+    SEED = 123
```
Comment (wjlee-ling) [inline corpus/twitter_classification/inference.py:54]: seed=42 가 저희, 허깅페이스 디폴트 설정입니다.

File: corpus/twitter_classification/tweet_classification_config.yaml [hunk 6]
```diff
@@ -0,0 +1,28 @@
...
+  pickle_to_csv_path: "./data/processed_data/twitter/twitter_0126.csv"
+  inference_save_path: "./data/processed_data/twitter/twitter_0126_pred.csv"
+  final_save_path: "./data/processed_data/twitter/final/0126.csv"
+
+model:
+  name_or_path: "./corpus/twitter_classification/saved_models/beomi/KcELECTRA-base/01-26-22-35/checkpoint-3500" # beomi/KcELECTRA-base #  monologg/koelectra-base-v3-discriminator
+  resume:
+
+tokenizer:
+  max_length: 256
```
Comment (wjlee-ling) [inline corpus/twitter_classification/tweet_classification_config.yaml:13]: 256이 적당할 까요? 최대 140자니까 토큰화 해도 200자 전후..?

File: inference.py [hunk 7]
```diff
@@ -25,8 +25,13 @@ def main(config):
...
-    elif "bart" in config.model.name_or_path:
-        print("🔥 bart")
+    elif (
+        "bart" in config.model.name_or_path
+        or "bart".upper() in config.model.name_or_path
+        or "t5" in config.model.name_or_path
+        or "t5".upper() in config.model.name_or_path
+    ):
+        print("🔥 Enc-Dec")
         tokenizer = PreTrainedTokenizerFast.from_pretrained(config.model.name_or_path)
```
Comment (wjlee-ling) [inline inference.py:35]: tokenizer에  새 토큰을 안 넣었는데 아래서 model.resize_token_embeddings가 들어가네요?

File: inference.py [hunk 8]
```diff
@@ -25,8 +25,13 @@ def main(config):
...
         model = AutoModelForSeq2SeqLM.from_pretrained(config.model.name_or_path)
```
Comment (wjlee-ling) [inline inference.py:36]: FlaxPretrainedModel로 사전훈련 된 모델도 이걸로 임포트가 될는지??

File: train.py [hunk 9]
```diff
@@ -28,15 +28,25 @@ def main(config):
...
         model = GPT2LMHeadModel.from_pretrained(config.model.name_or_path)
         model.resize_token_embeddings(len(tokenizer))
-    elif "bart" in config.model.name_or_path:
+    elif (
+        "bart" in config.model.name_or_path
+        or "bart".upper() in config.model.name_or_path
+        or "t5" in config.model.name_or_path
+        or "t5".upper() in config.model.name_or_path
+    ):
         model = AutoModelForSeq2SeqLM.from_pretrained(config.model.name_or_path)
```
Comment (wjlee-ling) [inline train.py:50]: 상동

File: trainer/trainer.py [hunk 10]
```diff
@@ -77,19 +78,35 @@ class BART_Chatbot:
...
-            data_collator=self.data_collator,
-            callbacks=[EarlyStoppingCallback(early_stopping_patience=self.config.callbacks.early_stopping_patience)],
-        )
+        self.test_dataset = self.datasets.raw_datasets["test"]
+
+        if self.config.train_mode == "pretraining":
+            if "bart" in self.config.model.name_or_path or "bart".upper() in self.config.model.name_or_path:
+                self.data_collator = DataCollatorForWholeWordMask(tokenizer=self.tokenizer, return_tensors="pt")
+            elif "t5" in self.config.model.name_or_path or "t5".upper() in self.config.model.name_or_path:
+                self.data_collator = DataCollatorForLanguageModeling(self.tokenizer, mlm=False, return_tensors="pt")
```
Comment (wjlee-ling) [inline trainer/trainer.py:87]: 이거 mlm=false 하면 clm으로 훈련하는거 아닌가요? 어차피  Flax 용 datacollator로 바뀔 예정인가요?

File: chatbot/retriever/dpr/dpr_dataset.py (hunk 1 above)
Comment (kimbyeolhee) [reply chatbot/retriever/dpr/dpr_dataset.py:24]: 네 이 부분이랑 이거 받아오는 부분 수정할게요!

File: chatbot/retriever/elastic_retriever.py (hunk 3 above)
Comment (kimbyeolhee) [reply chatbot/retriever/elastic_retriever.py:40]: 그거 그냥 이름을 v1이름으로 지어놓은거에요! 정보성 데이터가 없어서 임시로 트위터 챗봇 데이터를 json으로 저장해놓고 불러오는 거에요!

File: corpus/twitter_classification/inference.py (hunk 5 above)
Comment (kimbyeolhee) [reply corpus/twitter_classification/inference.py:54]: 넵 42로 수정하겠습니다

File: corpus/twitter_classification/tweet_classification_config.yaml (hunk 6 above)
Comment (kimbyeolhee) [reply corpus/twitter_classification/tweet_classification_config.yaml:13]: 보통 128, 256 이렇게 2의 거듭제곱수로 입력하길래 140자여서 128보다는 크게 해야할 것 같아가지고 256으로 설정해놓았는데 어떤 숫자가 적합할까요?

Comment (wjlee-ling) [reply corpus/twitter_classification/tweet_classification_config.yaml:13]: 물론 크게 잡으면 안 잘릴텐데, 크게 잡으면 그만큼 코스트가 늘어나고 생성 속도도 영점 몇초지만 늘어날 것 같아서... 그럼 256으로 가시죠

File: inference.py (hunk 7 above)
Comment (wjlee-ling) [reply inference.py:35]: 제가 어제 FlaxT5 구조로 모델 임포트 했을 때 model.resize_token_embeddings가 안 먹혔어요. 그런 attribute이 없다고... 허깅페이스의 config에서 바꿔 준 상태에서 임포트해야 하더라고요.

Comment (ezez-refer) [reply inference.py:35]: 현재는 새 토큰이 추가 되지 않았고 추후에 추가할지 몰라서 넣어뒀어요.

File: chatbot/retriever/dpr/dpr_dataset.py (hunk 1 above)
Comment (kimbyeolhee) [reply chatbot/retriever/dpr/dpr_dataset.py:24]: 아 이 부분이 query랑 context에 대해서 각각 input_ids랑 token_type_ids, attention mask를 받아야해서 데이터로더에서는 저렇게 보내고
모델에 넣을 때는 딕셔너리 형태로 각각 넣어줘요!

Comment (kimbyeolhee) [reply chatbot/retriever/dpr/dpr_dataset.py:24]: https://github.com/boostcampaitech4lv23nlp2/final-project-level3-nlp-13/pull/17/files#diff-03ef86420c743664dcdad2933ea8e55add53764f11826bc2ab601113d92a6d61R61

File: corpus/twitter_classification/inference.py (hunk 5 above)
Comment (kimbyeolhee) [reply corpus/twitter_classification/inference.py:54]: ✅ 수정완료

Comment (wjlee-ling) [reply corpus/twitter_classification/inference.py:54]: 여기 말고 또 있었던거 같은데... 어디였는지 기억이 안나네요. 죄송해요 봤을 때 체크했어야 했는데

File: trainer/trainer.py (hunk 10 above)
Comment (ezez-refer) [reply trainer/trainer.py:87]: 아뇨 finetuning은 DataCollatorForLanguageModeling(mlm=False)로 그대로 쓸 예정입니다

File: inference.py (hunk 8 above)
Comment (ezez-refer) [reply inference.py:36]: 네 from_flax=True 설정으로 불러올 수 있어요

File: train.py (hunk 9 above)
Comment (ezez-refer) [reply train.py:50]: 상동

File: corpus/twitter_classification/inference.py (hunk 5 above)
Comment (kimbyeolhee) [reply corpus/twitter_classification/inference.py:54]: 네네 그 train 부분도 123이었는데 수정했어요!
